nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,159 @@ defmodule Nx.BinaryBackend do
     from_binary(out, output_data)
   end
 
+  @impl true
+  def qr(
+        {%{shape: {m, k} = shape_q} = q_holder,
+         %{shape: {k, n} = shape_r, type: {_, num_bits} = type} = r_holder},
+        tensor,
+        opts
+      ) do
+    mode = opts[:mode]
+
+    r =
+      if mode == :reduced do
+        tensor[[0..(k - 1), 0..(n - 1)]]
+      else
+        tensor
+      end
+      |> Nx.as_type(type)
+      |> to_binary()
+
+    shape_h = {k, k}
+
+    {q_bin, r_bin} =
+      for i <- 0..(n - 1), reduce: {nil, r} do
+        {q, r} ->
+          # a = r[[i..(k - 1), i]]
+          {a, num_rows_a, _, _} =
+            for <<x::bitstring-size(num_bits) <- r>>, reduce: {<<>>, 0, 0, 0} do
+              {data, num_rows_a_prev, row, col} ->
+                {item, num_rows_a} =
+                  if row >= i and col == i do
+                    {x, num_rows_a_prev + 1}
+                  else
+                    {"", num_rows_a_prev}
+                  end
+
+                {row, col} =
+                  if col + 1 == n do
+                    {row + 1, 0}
+                  else
+                    {row, col + 1}
+                  end
+
+                {data <> item, num_rows_a, row, col}
+            end
+
+          h_bin = householder_reflector(a, num_rows_a, type, k)
```

Comment:
I think it will be simpler if you build a list of values here and then work with lists in householder_reflector. Something like this:

```suggestion
          {a, num_rows_a, _, _} =
            match_types [input_type] do
              for <<match!(x, 0) <- r>>, reduce: {[], 0, 0, 0} do
                {acc, num_rows_a_prev, row, col} ->
                  {acc, num_rows_a} =
                    if row >= i and col == i do
                      {[read!(x, 0) | acc], num_rows_a_prev + 1}
                    else
                      {acc, num_rows_a_prev}
                    end

                  {row, col} =
                    if col + 1 == n do
                      {row + 1, 0}
                    else
                      {row, col + 1}
                    end

                  {acc, num_rows_a, row, col}
              end
            end

          h_bin = householder_reflector(Enum.reverse(a), num_rows_a, type, k)
```

It will also be more performant because you will avoid converting to number multiple times in the household function.